internal/codeintel/dependencies/service.go

```diff
@@ -174,34 +173,94 @@ func (s *Service) resolveRepoCommits(ctx context.Context, repoRevs map[api.RepoN
 	return resolvedCommits, nil
 }
 
-// lockfileDependencies returns a flattened list of package dependencies for every repo and
-// revision pair specified in the given map.
-func (s *Service) lockfileDependencies(ctx context.Context, repoCommits []repoCommitResolvedCommit) ([]reposource.PackageDependency, error) {
-	var (
-		mu   sync.Mutex
-		deps []reposource.PackageDependency
-	)
+// lockfileDependencies returns a flattened list of package dependencies for every repo-commit pair.
+func (s *Service) lockfileDependencies(ctx context.Context, repoCommits []repoCommitResolvedCommit) (deps []shared.PackageDependency, _ error) {
+	// resolverFunc describes internal functions that perform bulk queries to gather the dependencies of
+	// some portion of the input. It is expected that if there are any unqueried repo-commit pairs remain
+	// that they are moved to the front of the given slice, and the number of unqueried elements returned.
+	type resolverFunc func(ctx context.Context, repoCommits []repoCommitResolvedCommit) ([]shared.PackageDependency, int, error)
+
+	resolvers := []resolverFunc{
+		s.resolveLockfileDependenciesFromStore,
+		s.resolveLockfileDependenciesFromArchive,
+	}
+
+	for _, resolver := range resolvers {
+		resolvedDeps, n, err := resolver(ctx, repoCommits)
+		if err != nil {
+			return nil, err
+		}
+
+		deps = append(deps, resolvedDeps...)
+		repoCommits = repoCommits[:n]
+	}
+
+	return deps, nil
+}
+
+// resolveLockfileDependenciesFromStore returns a flattened list of package dependencies for each
+// of the given repo-commit pairs from the database. The given `repoCommits` slice is altered in-place.
+// The returned `numUnqueried` value is the number of elements at the prefix of the slice that had no data.
+// It is expected that the remaining elements be passed to the fallback dependencies resolver, if one is
+// registered.
+func (s *Service) resolveLockfileDependenciesFromStore(ctx context.Context, repoCommits []repoCommitResolvedCommit) (deps []shared.PackageDependency, numUnqueried int, err error) {
+	ctx, _, endObservation := s.operations.resolveLockfileDependenciesFromStore.With(ctx, &err, observation.Args{})
+	defer func() {
+		endObservation(1, observation.Args{LogFields: []log.Field{
+			log.Int("numUnqueried", numUnqueried),
+		}})
+	}()
+
+	// Filter in-place
+	unqueried := repoCommits[:0]
+
+	for _, repoCommit := range repoCommits {
+		// TODO - batch these requests in the store layer
+		repoDeps, err := s.dependenciesStore.LockfileDependencies(ctx, string(repoCommit.Repo), repoCommit.ResolvedCommit)
+		if err != nil {
+			return nil, 0, errors.Wrap(err, "store.LockfileDependencies")
+		}
+		if len(repoDeps) > 0 {
+			deps = append(deps, repoDeps...)
+		} else {
+			unqueried = append(unqueried, repoCommit)
+		}
+	}
+
+	return nil, len(unqueried), nil
```

Comment:
I think it has to be this:

```suggestion
	return deps, len(unqueried), nil
```

Otherwise this always returns a nil slice and the dependencies will be added again.

Reply:
YES. Good catch thank you! (applied this change + some others locally, will push soon)